src/programs/city/defense.js

```diff
@@ -12,10 +12,72 @@ class CityDefense extends kernel.process {
       return this.suicide()
     }
 
-    this.safeMode()
+    const room = Game.rooms[this.data.room]
+
+    let towers = sos.lib.cache.getOrUpdate(
+      [this.data.room, 'towers'],
+      { persist: true, maxttl: 5000, chance: 0.01 },
+      () => room.find(FIND_MY_STRUCTURES, s => s.structureType === STRUCTURE_TOWER))
+
+    const hostiles = room.find(FIND_HOSTILE_CREEPS)
+
+    if (towers && towers.length > 0) {
+      this.fireTowers(towers, hostiles)
+    }
+
+    if (hostiles.some(c => c.owner.username === 'Invader')) {
```

Comment:
Shouldn't this check that the names *aren't* `Invader`? Unless I'm missing something this will only trigger safemode for invaders, as opposed to only triggering it for anything but invaders.

Reply:
ah, excellent catch. Fixed (and improved, even)